server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.util;
+
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+public abstract class TopicFilter {
```

Comment:
We cannot replicate the sealed behavior until Java 17, but we can have a single file containing all implementations like this:

```sh
public interface TopicFilter {
    boolean isTopicAllowed(String topic, boolean excludeInternalTopics);
    
    public static class IncludeList implements TopicFilter {
      // ...
    }
}
```

Reply:
I had thought about this at the time of implementation. The reason I didn't do this was that the original implementation used `sealed abstract class` which probably helps in exhaustive pattern matching. Also, I am not sure how would having the classes in the same file help in this case. Maybe I am missing something?

Reply:
I helps with reducing the number of source files and clarity as you have all implementations on the same page. This is similar to what the original scala code and TopicPartitionFilter are doing.